File: Python/ArraysAndHashing/GroupAnagrams.py
def groupAnagrams(strs: list[str]) -> list[list[str]]:
    groups = {}

    for word in strs:
        count = [0] * 26

        for ch in word:
            count[ord(ch) - ord('a')] += 1

        signature = tuple(count)

        if signature not in groups:
            groups[signature] = []

        groups[signature].append(word)

    return list(groups.values())

File: TypeScript/ArraysAndHashing/GroupAnagrams.ts
function groupAnagrams(strs: string[]): string[][] {
    const groups = new Map<string, string[]>();

    for (const word of strs) {
        const count = new Array(26).fill(0);

        for (const ch of word) {
            const index = ch.charCodeAt(0) - 'a'.charCodeAt(0);
            count[index]++;
        }

        const signature = count.join("#");

        if (!groups.has(signature)) {
            groups.set(signature, []);
        }

        groups.get(signature)!.push(word);
    }

    return Array.from(groups.values());
}
